snapfaas/src/fs/mod.rs

```diff
@@ -893,6 +959,29 @@ impl<S: BackingStore> FS<S> {
         })
     }
 
+    pub fn invoke_service(&self, service: &Service) -> Result<ServiceInfo, ServiceError> {
+        CURRENT_LABEL.with(|current_label| {
+            if current_label
+                .borrow()
+                .integrity
+                .implies(&service.weakest_privilege_required)
+               && current_label
+                .borrow()
+                .can_flow_to_with_privilege(&Buckle::public(), &service.privilege)
```

Comment:
@alevy My understanding is that both ways are secure---either we just check can_flow_to wrt the privilege or we actually declassify lowering the label and then check can_flow_to.

The two differs in what label the function has upon return. If permitted, the former has `lub(old_label, ret_val_label)`, the latter has the `lub(pub, ret_val_label)`. If not permitted, the former has the same `old_label`, the latter has the lowest label the function can declassify to.

I think we should actually declassify as it is what a function would do?

Reply:
Hrm... the function should not have it's label lowered just because it called a gate (this would declassify other things the function has read)

Reply:
Right. I meant if service were implemented as a function, the function would need to call declassify to lower its label.

I understand `can_flow_to_with_privilege` as a shortcut to
```
old_label = current_label()
declassify_to_public()
invoke_service(...)
taint_with_label(old_label)
```

Reply:
If service were implemented as a function, I suppose the declassification happens within the service function, thus the invoker's label won't change.

Reply:
@tan-yue yes, I agree. In other words, the (function-as-a-? heh? get it?)service would declassify to public (so i can send packets out on the network). Assuming it can, it would then perform the request, tainting itself with the label it decides is appropriate for the return value and also tainting itself with its original label. Thus, the result is that the service needs to be able to declassify the caller's label to public and the caller's label should only increase to the least-upper-bound of it's label before invoking the service and the return values label.

Reply:
> If service were implemented as a function, I suppose the declassification happens within the service function, thus the invoker's label won't change.

You are right. I failed to realize `current_label` is the invoker's label.

I have no more question.